### ISSUE TITLE
runtime error: slice bounds out of range

### ISSUE THREAD

**Issue by janstrunk** [weaviate/weaviate#5021]
### How to reproduce this bug?

We ran into "runtime error: slice bounds out of range" when we upgraded a self-hosted Weaviate server instance from version 1.19.8 to version 1.24.8.

Just as a background, here are some infos on what our database base looks like.

**Structure**
In our database, we have four data classes: Chunk, Document, Corpus, and Task.
Corpora contain Documents, which is stored by the documents having a corpus property that cross-references the Corpus object they belong to. Likewise, Documents contain Chunks, which is again stored by chunks having a document property that cross-references the Document object they belong to.

Cross-reference scheme: Chunk → Document → Corpus

**Content**
Before the update, our database contained:
220 Corpus objects
44,722 Document objects
2,283,304 Chunk objects
242 Task objects

After the update of the Weaviate server to version 1.24.8, when our RAG system tried to use Aggregate queries to count, e.g. the number of documents in a corpus or the number of chunks in a corpus. We encountered the following error:
```
{"level":"error","msg":"Recovered from panic: runtime error: slice bounds out of range [:44714] with capacity 10000, local variables [[{1 0xc03af2d530 0xc03b0b9e20 []}]], additional lo │
│ goroutine 178304 [running]:                                                                                                                                                              │
│ runtime/debug.Stack()                                                                                                                                                                    │
│     /usr/local/go/src/runtime/debug/stack.go:24 +0x5e                                                                                                                                    │
│ runtime/debug.PrintStack()                                                                                                                                                               │
│     /usr/local/go/src/runtime/debug/stack.go:16 +0x13                                                                                                                                    │
│ github.com/weaviate/weaviate/entities/errors.(*ErrorGroupWrapper).setDeferFunc.func1({0xc0034c03f0, 0x1, 0x1})                                                                           │
│     /go/src/github.com/weaviate/weaviate/entities/errors/error_group_wrapper.go:74 +0x145                                                                                                │
│ panic({0x1953a60?, 0xc00306c768?})                                                                                                                                                       │
│     /usr/local/go/src/runtime/panic.go:914 +0x21f                                                                                                                                        │
│ github.com/weaviate/weaviate/adapters/repos/db.(*DB).getStoreObjectsWithScores(...)                                                                                                      │
│     /go/src/github.com/weaviate/weaviate/adapters/repos/db/search.go:420                                                                                                                 │
│ github.com/weaviate/weaviate/adapters/repos/db.(*DB).Search(_, {_, _}, {0xc00221f0b8, {0xc03a4676e0, 0x8}, 0xc00306c750, 0x0, {0x0, 0x0, ...}, ...})                                     │
│     /go/src/github.com/weaviate/weaviate/adapters/repos/db/search.go:99 +0x287                                                                                                           │
│ github.com/weaviate/weaviate/adapters/repos/db/inverted.(*refFilterExtractor).fetchIDs(0xc00706dd00, {0x1e7c730, 0x2b83dc0})                                                             │
│     /go/src/github.com/weaviate/weaviate/adapters/repos/db/inverted/searcher_ref_filter.go:121 +0x1a8                                                                                    │
│ github.com/weaviate/weaviate/adapters/repos/db/inverted.(*refFilterExtractor).Do(0xc004607d00, {0x1e7c730, 0x2b83dc0})                                                                   │
│     /go/src/github.com/weaviate/weaviate/adapters/repos/db/inverted/searcher_ref_filter.go:66 +0x66                                                                                      │
│ github.com/weaviate/weaviate/adapters/repos/db/inverted.(*Searcher).extractReferenceFilter(0xc00706ddac?, 0xc03a466f78?, 0xc003646600?, 0xc03a4676d8?)                                   │
│     /go/src/github.com/weaviate/weaviate/adapters/repos/db/inverted/searcher.go:312 +0x9d                                                                                                │
│ github.com/weaviate/weaviate/adapters/repos/db/inverted.(*Searcher).extractPropValuePair(0xc03b13ed00, 0xc03af2d7d0, {0xc03a466f78, 0x5})                                                │
│     /go/src/github.com/weaviate/weaviate/adapters/repos/db/inverted/searcher.go:254 +0x445                                                                                               │
│ github.com/weaviate/weaviate/adapters/repos/db/inverted.(*Searcher).extractPropValuePairs.func1()                                                                                        │
│     /go/src/github.com/weaviate/weaviate/adapters/repos/db/inverted/searcher.go:292 +0x45                                                                                                │
```

This error then made it impossible to create new Document or Chunk objects since our system updates the statistics for the Corpus object that contains these documents and chunks using Aggregate queries. We found no way of changing the configuration to remedy this problem.

Since we thought that the update may have corrupted something in our database, we dumped the contents of our database and restarted the database from scratch, deleting all schemas.
While restoring the data from the dump bit by bit, we noticed that with version 1.24.8 of Weaviate, there seemed to be a certain threshold of the number of objects before which Aggregate queries worked fine and inserting new objects was possible. But after crossing that threshold, we again encountered the "slice bounds out of range" error from above. However, if we then deleted the last inserted objects again going back under the threshold of the number of objects in the database, performing Aggregate queries (and inserting new objects again) worked without an error (even without restarting the Weaviate server in between).

**Problem threshold**
With
148 Corpus objects
10,002 Document objects
346,874 Chunk objects
there was no "slice bounds out of range" error when doing Aggregate queries (and inserting new objects was possible without problem).

However, after adding
+ 2 Corpus objects
+ 212 Document objects
+ 4,300 Chunk objects
we again encountered the same problems described above. When we removed these newly added objects again, everything was fine again even without restarting the Weaviate server in between.

**Temporary solution for us**
The interesting thing is that once we downgraded our Weaviate server to version 1.24.3, these problems disappeared. That is, we could increase the number of objects substantially without encountering any "slice bounds out of range" errors
during Aggregate queries and we could add more and more objects without any problems. (This is in line with the fact that, before our update, we had a Weaviate server with version 1.19.8 running fine with many more objects in it without any problems...)

**Summary**
Therefore, we suspect that some kind of bug has been introduced between version 1.24.3 and 1.24.8 of the Weaviate server that causes problems with (Aggregate?) queries with databases with many objects and we would be very thankful if you could investigate this and we would of course also be ready to help with further information if needed. The problem seems to be similar to an earlier closed issue: https://github.com/weaviate/weaviate/issues/1897

Thank you very much! 

### What is the expected behavior?

**Temporary solution for us**
The interesting thing is that once we downgraded our Weaviate server to version 1.24.3, these problems disappeared. That is, we could increase the number of objects substantially without encountering any "slice bounds out of range" errors
during Aggregate queries and we could add more and more objects without any problems. (This is in line with the fact that, before our update, we had a Weaviate server with version 1.19.8 running fine with many more objects in it without any problems...)

### What is the actual behavior?

With Weaviate server version 1.24.8, we got the following error whenever we wanted to run Aggregate queries on our database once we crossed a certain threshold of the number of Corpus, Document, and Chunk objects in the database (see below).

```
{"level":"error","msg":"Recovered from panic: runtime error: slice bounds out of range [:44714] with capacity 10000, local variables [[{1 0xc03af2d530 0xc03b0b9e20 []}]], additional lo │
│ goroutine 178304 [running]:                                                                                                                                                              │
│ runtime/debug.Stack()                                                                                                                                                                    │
│     /usr/local/go/src/runtime/debug/stack.go:24 +0x5e                                                                                                                                    │
│ runtime/debug.PrintStack()                                                                                                                                                               │
│     /usr/local/go/src/runtime/debug/stack.go:16 +0x13                                                                                                                                    │
│ github.com/weaviate/weaviate/entities/errors.(*ErrorGroupWrapper).setDeferFunc.func1({0xc0034c03f0, 0x1, 0x1})                                                                           │
│     /go/src/github.com/weaviate/weaviate/entities/errors/error_group_wrapper.go:74 +0x145                                                                                                │
│ panic({0x1953a60?, 0xc00306c768?})                                                                                                                                                       │
│     /usr/local/go/src/runtime/panic.go:914 +0x21f                                                                                                                                        │
│ github.com/weaviate/weaviate/adapters/repos/db.(*DB).getStoreObjectsWithScores(...)                                                                                                      │
│     /go/src/github.com/weaviate/weaviate/adapters/repos/db/search.go:420                                                                                                                 │
│ github.com/weaviate/weaviate/adapters/repos/db.(*DB).Search(_, {_, _}, {0xc00221f0b8, {0xc03a4676e0, 0x8}, 0xc00306c750, 0x0, {0x0, 0x0, ...}, ...})                                     │
│     /go/src/github.com/weaviate/weaviate/adapters/repos/db/search.go:99 +0x287                                                                                                           │
│ github.com/weaviate/weaviate/adapters/repos/db/inverted.(*refFilterExtractor).fetchIDs(0xc00706dd00, {0x1e7c730, 0x2b83dc0})                                                             │
│     /go/src/github.com/weaviate/weaviate/adapters/repos/db/inverted/searcher_ref_filter.go:121 +0x1a8                                                                                    │
│ github.com/weaviate/weaviate/adapters/repos/db/inverted.(*refFilterExtractor).Do(0xc004607d00, {0x1e7c730, 0x2b83dc0})                                                                   │
│     /go/src/github.com/weaviate/weaviate/adapters/repos/db/inverted/searcher_ref_filter.go:66 +0x66                                                                                      │
│ github.com/weaviate/weaviate/adapters/repos/db/inverted.(*Searcher).extractReferenceFilter(0xc00706ddac?, 0xc03a466f78?, 0xc003646600?, 0xc03a4676d8?)                                   │
│     /go/src/github.com/weaviate/weaviate/adapters/repos/db/inverted/searcher.go:312 +0x9d                                                                                                │
│ github.com/weaviate/weaviate/adapters/repos/db/inverted.(*Searcher).extractPropValuePair(0xc03b13ed00, 0xc03af2d7d0, {0xc03a466f78, 0x5})                                                │
│     /go/src/github.com/weaviate/weaviate/adapters/repos/db/inverted/searcher.go:254 +0x445                                                                                               │
│ github.com/weaviate/weaviate/adapters/repos/db/inverted.(*Searcher).extractPropValuePairs.func1()                                                                                        │
│     /go/src/github.com/weaviate/weaviate/adapters/repos/db/inverted/searcher.go:292 +0x45                                                                                                │
```

**Problem threshold**
With
148 Corpus objects
10,002 Document objects
346,874 Chunk objects
there was no "slice bounds out of range" error when doing Aggregate queries (and inserting new objects was possible without problem).

However, after adding
+ 2 Corpus objects
+ 212 Document objects
+ 4,300 Chunk objects
we again encountered the same problems described above. When we removed these newly added objects again, everything was fine again even without restarting the Weaviate server in between.

### Supporting information

This bug seems to be similar to earlier but now closed issue: https://github.com/weaviate/weaviate/issues/1897

### Server Version

1.24.8

### Code of Conduct

- [X] I have read and agree to the Weaviate's [Contributor Guide](https://weaviate.io/developers/contributor-guide) and [Code of Conduct](https://weaviate.io/service/code-of-conduct)

**Comment by evenfrost** [comment]
Having the same issue with `v1.24.12`, downgrading to `1.22.6` helped.

The error text is:
```
slice bounds out of range [:15822] with capacity 10000
```

**Comment by janstrunk** [comment]
The issue also occurs with version 1.24.6. I tested that, too.

**Comment by dirkkul** [comment]
Hi, could you please test `semitechnologies/weaviate:preview-fix-score-extraction-5e22dcc` and confirm that it fixes the issue?

**Comment by dirkkul** [comment]
fixed by https://github.com/weaviate/weaviate/pull/5066, should be in the next 1.24/25 point release

**Comment by janstrunk** [comment]
> Hi, could you please test `semitechnologies/weaviate:preview-fix-score-extraction-5e22dcc` and confirm that it fixes the issue? (It is based on the latest 1.24 release)

Is there an easy way to deploy this specific branch on Kubernetes using Helm?
If not, I will definitely test the next 1.24/25 point release with our existing data.

Thank you very much for the quick work!

**Comment by janstrunk** [comment]
> > Hi, could you please test `semitechnologies/weaviate:preview-fix-score-extraction-5e22dcc` and confirm that it fixes the issue? (It is based on the latest 1.24 release)
> 
> @dirkkul Is there an easy way to deploy this specific branch on Kubernetes using Helm? If not, I will definitely test the next 1.24/25 point release with our existing data.
@dirkkul Ah, ok, I could probably deploy using the tag you gave above using Helm. So, I'll try that with our data as soon as possible...

**Comment by dirkkul** [comment]
I sadly don't really know how Helm works, so I cannot help you there
